sdk/core/core-rest-pipeline/src/policies/retryPolicy.ts

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { PipelineResponse, PipelineRequest, SendRequest } from "../interfaces";
+import { PipelinePolicy } from "../pipeline";
+import { delay } from "../util/helpers";
+import { createClientLogger } from "@azure/logger";
+import { RetryStrategy, RetryInformation, RetryModifiers } from "../retryStrategies/retryStrategy";
+import { RestError } from "../restError";
+import { AbortError } from "@azure/abort-controller";
+
+const retryPolicyLogger = createClientLogger("core-rest-pipeline retryPolicy");
+const DEFAULT_MAX_RETRIES = 3;
+
+/**
+ * The programmatic identifier of the retryPolicy.
+ */
+const retryPolicyName = "retryPolicy";
+
+/**
+ * Options to the {@link retryPolicy}
+ */
+export interface RetryPolicyOptions {
+  /**
+   * Maximum number of retries. If not specified, it will limit to 3 retries.
+   */
+  maxRetries: number;
+}
+
+/**
+ * retryPolicy is a generic policy to enable retrying requests when certain conditions are met
+ */
+export function retryPolicy(
+  strategies: RetryStrategy[],
+  options: RetryPolicyOptions = { maxRetries: DEFAULT_MAX_RETRIES }
+): PipelinePolicy {
+  return {
+    name: retryPolicyName,
+    async sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse> {
+      let response: PipelineResponse | undefined;
+      let responseError: RestError | undefined;
+      let retryCount = -1;
+
+      // eslint-disable-next-line no-constant-condition
+      retryRequest: while (true) {
+        retryCount += 1;
+        response = undefined;
+        responseError = undefined;
+
+        try {
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Attempting to send request`,
+            request.requestId
+          );
+          response = await next(request);
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Received a response from request`,
+            request.requestId
+          );
+        } catch (e) {
+          retryPolicyLogger.info(
+            `Retry ${retryCount}: Received an error from request`,
+            request.requestId
+          );
+          responseError = e as RestError;
+          if (responseError.name !== "RestError") {
+            throw responseError;
+          }
+          response = responseError.response;
+        }
+
```

Comment:
This is a problem because, if a retry strategy sets the `retryAfterInMs` property, this change means that the last retry would only return waiting the milliseconds specified through `retryAfterInMs`.

We can keep this change, but to make it work, I’ll move the processing of `retryAfterInMs` to be only after this code here.

Reply:
Actually, on second thought, I’ll undo this change. Let’s discuss it tomorrow!

Reply:
Ok what made more sense for me was to keep this change before the processing of strategies but after sending the request.

Reply:
Updated here: https://github.com/Azure/azure-sdk-for-js/pull/19078/commits/7f84291de33c9137f9c0257251815095c004b9e1